src/wp-includes/html-api/class-wp-html-doctype-info.php

```diff
@@ -0,0 +1,473 @@
+		}
+
+		$this->compatibility_mode = 'quirks';
+		return;
+
+		parse_doctype_public_identifier:
+		/*
+		 * The parser should enter "DOCTYPE public identifier (double-quoted) state" or
+		 * "DOCTYPE public identifier (single-quoted) state" by finding one of the valid quotes.
+		 * Anything else forces quirks mode and ignores the rest of the contents.
+		 *
+		 * @see https://html.spec.whatwg.org/multipage/parsing.html#doctype-public-identifier-(double-quoted)-state
+		 * @see https://html.spec.whatwg.org/multipage/parsing.html#doctype-public-identifier-(single-quoted)-state
+		 */
+		$closer_quote = $doctype_content[ $at ];
+		if ( '"' !== $closer_quote && "'" !== $closer_quote ) {
+			$this->compatibility_mode = 'quirks';
+			return;
+		}
+		++$at;
+
+		$identifier_length       = strcspn( $doctype_content, $closer_quote, $at, $end - $at );
+		$this->public_identifier = substr( $doctype_content, $at, $identifier_length );
+
+		$at += $identifier_length;
+		if ( $at >= $end || $closer_quote !== $doctype_content[ $at ] ) {
+			$this->compatibility_mode = 'quirks';
+			return;
+		}
+		++$at;
+
+		/*
+		 * "Between DOCTYPE public and system identifiers state"
+		 *
+		 * Advance through whitespace between public and system identifiers.
+		 *
+		 * @see https://html.spec.whatwg.org/multipage/parsing.html#between-doctype-public-and-system-identifiers-state
+		 */
+		$at += strspn( $doctype_content, " \t\n\f\r", $at, $end - $at );
+		if ( $at >= $end ) {
+			$this->compatibility_mode = self::apply_quirks_mode_algorithm( $this->name, $this->public_identifier, $this->system_identifier );
+			return;
+		}
+
+		parse_doctype_system_identifier:
+		/*
+		 * The parser should enter "DOCTYPE system identifier (double-quoted) state" or
+		 * "DOCTYPE system identifier (single-quoted) state" by finding one of the valid quotes.
+		 * Anything else forces quirks mode and ignores the rest of the contents.
+		 *
+		 * @see https://html.spec.whatwg.org/multipage/parsing.html#doctype-system-identifier-(double-quoted)-state
+		 * @see https://html.spec.whatwg.org/multipage/parsing.html#doctype-system-identifier-(single-quoted)-state
+		 */
+		$closer_quote = $doctype_content[ $at ];
+		if ( '"' !== $closer_quote && "'" !== $closer_quote ) {
+			$this->compatibility_mode = 'quirks';
+			return;
+		}
+		++$at;
+
+		$identifier_length       = strcspn( $doctype_content, $closer_quote, $at, $end - $at );
+		$this->system_identifier = substr( $doctype_content, $at, $identifier_length );
+
+		$at += $identifier_length;
+		if ( $at >= $end || $closer_quote !== $doctype_content[ $at ] ) {
+			$this->compatibility_mode = 'quirks';
+			return;
+		}
+
+		$this->compatibility_mode = self::apply_quirks_mode_algorithm( $this->name, $this->public_identifier, $this->system_identifier );
+	}
+
+	/**
+	 * Gets the name of the DOCTYPE.
+	 *
+	 * @since 6.7.0
+	 *
+	 * @return string The name of the DOCTYPE.
+	 */
+	public function get_name(): string {
```

Comment:
I almost think it would be more valuable to represent the absence of these identifiers, just as `get_attribute()` does. the doctype declaration can have an empty name and it can lack a name, I guess. maybe this is more about the system and public identifier.

if I were writing a method to copy the token from a source document into a normalized output, however, I would want to know the difference.

I also find it unlikely that userspace code is going to be doing much more with this than possibly copying HTML. I would like to see what people want or expect to do with the information, but it's not like we see people parsing DOCTYPE tokens already using the existing tools available